flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/FFT.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.utils;
+
+import org.apache.commons.math3.complex.Complex;
+
+/**
+ * Fast Fourier Transformation(FFT).
+ * Provides 2 algorithms:
+ * 1. Cooley-Tukey algorithm, high performance, but only supports length of power-of-2.
+ * 2. Chirp-Z algorithm, can perform FFT with any length.
+ */
+public class FFT {
+	/**
+	 * Helper for root of unity. Returns the group for power "length".
+	 */
+	public static Complex[] getOmega(int length) throws Exception {
+		Complex[] omega = new Complex[length];
+		Complex unit = new Complex(Math.cos(2 * Math.PI / length),
+			Math.sin(2 * Math.PI / length));
+		omega[0] = new Complex(1, 0);
+		omega[1] = unit;
+		for (int index = 2; index < length; index++) {
+			omega[index] = omega[index - 1].multiply(unit);
+		}
+		return omega;
+	}
+
+	/**
+	 * Cooley-Tukey algorithm, can perform fft for any composite base.
+	 * Specifically, it can perform power-of-2 base fft(with some modification to make it an in-place algorithm).
+	 * See:
+	 * "An algorithm for the machine calculation of complex Fourier series", JW Cooley, JW Tukey, 1965
+	 * for a rough reference.
+	 * Detail of radix-2 in-place Cooley-Tukey algorithm can be found in many places, e.g. CLRS textbook.
+	 */
+	public static Complex[] fftRadix2CooleyTukey(Complex[] input, Boolean inverse, Complex[] omega) throws Exception {
```

Comment:
It seems that you never use "inverse" as "Boolean" which can be null, but only treat it as a simple boolean. It's better to avoid using boxed type unless it's strictly required.

By the way, in flink code style, all input params are "#NotNull" by default. It's recommended to add a "#Nullable" annotation for those who can be null. Please replace # with @, I use # because @ in github has specific usage.

Reply:
Thanks, "Boolean" is changed to "boolean"